compiler/reader.go

```diff
@@ -47,60 +50,94 @@ func initializeInfoCache() {
 	}
 }
 
+// EnableFileCache turns on file caching.
 func EnableFileCache() {
+	fileCacheMutex.Lock()
+	defer fileCacheMutex.Unlock()
 	fileCacheEnable = true
 }
 
+// EnableInfoCache turns on parsed info caching.
 func EnableInfoCache() {
+	infoCacheMutex.Lock()
+	defer infoCacheMutex.Unlock()
 	infoCacheEnable = true
 }
 
+// DisableFileCache turns off file caching.
 func DisableFileCache() {
+	fileCacheMutex.Lock()
+	defer fileCacheMutex.Unlock()
 	fileCacheEnable = false
 }
 
+// DisableInfoCache turns off parsed info caching.
 func DisableInfoCache() {
+	infoCacheMutex.Lock()
+	defer infoCacheMutex.Unlock()
 	infoCacheEnable = false
 }
 
+// RemoveFromFileCache removes an entry from the file cache.
 func RemoveFromFileCache(fileurl string) {
+	fileCacheMutex.Lock()
+	defer fileCacheMutex.Unlock()
 	if !fileCacheEnable {
 		return
 	}
 	initializeFileCache()
 	delete(fileCache, fileurl)
 }
 
+// RemoveFromInfoCache removes an entry from the info cache.
 func RemoveFromInfoCache(filename string) {
+	infoCacheMutex.Lock()
+	defer infoCacheMutex.Unlock()
 	if !infoCacheEnable {
 		return
 	}
 	initializeInfoCache()
 	delete(infoCache, filename)
 }
 
+// GetInfoCache returns the info cache map.
 func GetInfoCache() map[string]interface{} {
+	infoCacheMutex.Lock()
+	defer infoCacheMutex.Unlock()
 	if infoCache == nil {
 		initializeInfoCache()
 	}
 	return infoCache
 }
 
+// ClearFileCache clears the file cache.
 func ClearFileCache() {
+	fileCacheMutex.Lock()
+	defer fileCacheMutex.Unlock()
 	fileCache = make(map[string][]byte, 0)
 }
 
+// ClearInfoCache clears the info cache.
 func ClearInfoCache() {
+	infoCacheMutex.Lock()
+	defer infoCacheMutex.Unlock()
 	infoCache = make(map[string]interface{})
 }
 
+// ClearCaches clears all caches.
 func ClearCaches() {
 	ClearFileCache()
 	ClearInfoCache()
 }
 
 // FetchFile gets a specified file from the local filesystem or a remote location.
 func FetchFile(fileurl string) ([]byte, error) {
+	fileCacheMutex.Lock()
+	defer fileCacheMutex.Unlock()
+	return fetchFile(fileurl)
+}
+
+func fetchFile(fileurl string) ([]byte, error) {
 	var bytes []byte
 	initializeFileCache()
```

Comment:
I know this isn't part of the diff, but should this `initializeFileCache()` call be synchronized? Or should the `initializeFileCache` function itself have lock calls in it? It _can_ mutate a global variable, so should it be synchronized?

Reply:
This call is unsynchronized because it is in a private function (see my comment at line 106 of the new version)